src/Storages/Kafka/StorageKafka2.cpp

```diff
@@ -0,0 +1,1291 @@
+        ops.push_back(zkutil::makeCreateRequest(topic_partition_path, "", zkutil::CreateMode::Persistent, ignore_if_exists));
+        ops.push_back(zkutil::makeCreateRequest(lock_file_path, kafka_settings->kafka_replica_name.value, zkutil::CreateMode::Ephemeral));
+    }
+    Coordination::Responses responses;
+
+    if (const auto code = keeper_to_use.tryMulti(ops, responses); code != Coordination::Error::ZOK)
+    {
+        if (code != Coordination::Error::ZNODEEXISTS)
+            zkutil::KeeperMultiException::check(code, ops, responses);
+
+        // Possible optimization: check the content of lock files, if we locked them, then we can clean them up and retry to lock them.
+        return std::nullopt;
+    }
+
+    // We have the locks, let's gather the information we needed
+    TopicPartitionLocks locks;
+    {
+        auto tp_it = topic_partitions.begin();
+        auto path_it = topic_partition_paths.begin();
+        for (; tp_it != topic_partitions.end(); ++tp_it, ++path_it)
+        {
+            using zkutil::EphemeralNodeHolder;
+            LockedTopicPartitionInfo lock_info{
+                EphemeralNodeHolder::existing(*path_it / lock_file_name, keeper_to_use),
+                getNumber(keeper_to_use, *path_it / commit_file_name),
+                getNumber(keeper_to_use, *path_it / intent_file_name)};
+
+            LOG_TRACE(
+                log,
+                "Locked topic partition: {}:{} at offset {} with intent size {}",
+                tp_it->topic,
+                tp_it->partition_id,
+                lock_info.committed_offset.value_or(0),
+                lock_info.intent_size.value_or(0));
+            locks.emplace(TopicPartition(*tp_it), std::move(lock_info));
+        }
+    }
+
+    return locks;
+}
+
+
+void StorageKafka2::saveCommittedOffset(zkutil::ZooKeeper & keeper_to_use, const TopicPartition & topic_partition)
+{
+    const auto partition_prefix = getTopicPartitionPath(topic_partition);
+    keeper_to_use.createOrUpdate(partition_prefix / commit_file_name, toString(topic_partition.offset), zkutil::CreateMode::Persistent);
+    // This is best effort, if it fails we will try to remove in the next round
+    keeper_to_use.tryRemove(partition_prefix / intent_file_name, -1);
+    LOG_TEST(
+        log, "Saved offset {} for topic-partition [{}:{}]", topic_partition.offset, topic_partition.topic, topic_partition.partition_id);
+}
+
+void StorageKafka2::saveIntent(zkutil::ZooKeeper & keeper_to_use, const TopicPartition & topic_partition, int64_t intent)
+{
+    LOG_TEST(
+        log,
+        "Saving intent of {} for topic-partition [{}:{}] at offset {}",
+        intent,
+        topic_partition.topic,
+        topic_partition.partition_id,
+        topic_partition.offset);
+    keeper_to_use.createOrUpdate(
+        getTopicPartitionPath(topic_partition) / intent_file_name, toString(intent), zkutil::CreateMode::Persistent);
+}
+
+
+StorageKafka2::PolledBatchInfo StorageKafka2::pollConsumer(
```

Comment:
Is BlocksListSource better than using KafkaSource? Kafka source could not be reused?

Reply:
I had quite some discussion about this with Nikolai. Introducing a new, special source might be a better solution, but it wouldn't be smart to reuse the original KafkaSource because of the so different way of how messages are polled. I think I will have to make this logic into a separate source to support direct reads, otherwise I don't think we can support direct reads (which is kind of a must have to pass experimental state).

As a consequence of everything above, I would like to avoid creating a special source for now and tackle this issue later in a separate PR (which probably would make direct reads possible).